include/rmm/mr/device/sam_headroom_resource_adaptor.hpp

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <rmm/cuda_device.hpp>
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/mr/device/device_memory_resource.hpp>
+#include <rmm/mr/device/system_memory_resource.hpp>
+#include <rmm/resource_ref.hpp>
+
+namespace rmm::mr {
+/**
+ * @addtogroup device_resource_adaptors
+ * @{
+ * @file
+ */
+/**
+ * @brief Resource that adapts system memory resource to allocate memory with a headroom.
+ *
+ * System allocated memory (SAM) can be migrated to the GPU, but is never migrated back the host. If
+ * GPU memory is over-subscribed, this can cause other CUDA calls to fail with out-of-memory errors.
+ * To work around this problem, when using a system memory resource, we reserve some GPU memory as
+ * headroom for other CUDA calls, and only conditionally set its preferred location to the GPU if
+ * the allocation would not eat into the headroom.
+ *
+ * Since doing this check on every allocation can be expensive, the caller may choose to use other
+ * allocators (e.g. `binning_memory_resource`) for small allocations, and use this allocator for
+ * large allocations only.
+ *
+ * @tparam Upstream Type of the upstream resource used for allocation/deallocation. Must be
+ *                  `system_memory_resource`.
+ */
+template <typename Upstream>
+class sam_headroom_resource_adaptor final : public device_memory_resource {
+ public:
+  /**
+   * @brief Construct a headroom adaptor using `upstream` to satisfy allocation requests.
+   *
+   * @param upstream The resource used for allocating/deallocating device memory. Must be
+   *                 `system_memory_resource`.
+   * @param headroom Size of the reserved GPU memory as headroom
+   */
+  explicit sam_headroom_resource_adaptor(Upstream* upstream, std::size_t headroom)
+    : upstream_{upstream}, headroom_{headroom}
+  {
+    static_assert(std::is_same_v<system_memory_resource, Upstream>,
```

Comment:
Seems like this could work with any prefetchable upstream, including a SAM MR that is already wrapped with another adaptor.

Reply:
This is specifically designed to address the current shortcomings of SAM, so I'm not sure if it's useful for other things (e.g. managed memory). You are right about the adaptors though, but I'm leaning towards being more strict to avoid misuse. What do you think?

Reply:
Yeah, OK. We can always loosen it if needed.